File: quant1x/learn/fpgrowth/fp_growth.cpp
#include "fp_growth.h"
#include "fp_growth_core.h"

namespace quant1x {

// FPGrowth<T> 实现
template <typename T>
FPGrowth<T>::FPGrowth(double min_support) 
    : core_(std::make_unique<FPGrowthCore>(min_support)) {}

template <typename T>
FPGrowth<T>::FPGrowth(size_t min_support_count) 
    : core_(std::make_unique<FPGrowthCore>(min_support_count)) {}

template <typename T>
FPGrowth<T>::~FPGrowth() = default;

template <typename T>
FPGrowth<T>::FPGrowth(FPGrowth&&) noexcept = default;

template <typename T>
FPGrowth<T>& FPGrowth<T>::operator=(FPGrowth&&) noexcept = default;

template <typename T>
std::vector<typename FPGrowth<T>::FrequentPattern> FPGrowth<T>::mine(const Transactions& transactions) {
    if (transactions.empty()) {
        return {};
    }

    // 1. 建立映射 T -> size_t
    std::unordered_map<T, size_t> item_to_id;
    std::vector<T> id_to_item;
    id_to_item.push_back(T()); // 0号ID保留

    FPGrowthCore::Transactions core_transactions;
    core_transactions.reserve(transactions.size());

    for (const auto& tx : transactions) {
        FPGrowthCore::Transaction core_tx;
        core_tx.reserve(tx.size());
        for (const auto& item : tx) {
            if (item_to_id.find(item) == item_to_id.end()) {
                size_t new_id = id_to_item.size();
                item_to_id[item] = new_id;
                id_to_item.push_back(item);
            }
            core_tx.push_back(item_to_id[item]);
        }
        core_transactions.push_back(std::move(core_tx));
    }

    // 2. 调用核心算法
    auto core_patterns = core_->mine(core_transactions);

    // 3. 映射回 T
    std::vector<FrequentPattern> patterns;
    patterns.reserve(core_patterns.size());

    for (const auto& p : core_patterns) {
        ItemSet itemset;
        itemset.reserve(p.first.size());
        for (size_t id : p.first) {
            itemset.push_back(id_to_item[id]);
        }
        patterns.emplace_back(std::move(itemset), p.second);
    }

    return patterns;
}

// 显式实例化
template class FPGrowth<std::string>;
template class FPGrowth<size_t>;
template class FPGrowth<int>;
template class FPGrowth<long>;

} // namespace quant1x

File: quant1x/learn/fpgrowth/fp_growth.go
package fpgrowth

// Pattern 泛型结果
type Pattern[T any] struct {
	Items   []T
	Support float64
}

// FPGrowth 泛型实现
type FPGrowth[T comparable] struct {
	core *FPGrowthCore
}

func New[T comparable](minSupport float64) *FPGrowth[T] {
	return &FPGrowth[T]{
		core: NewCore(minSupport),
	}
}

func NewWithCount[T comparable](minSupportCount int) *FPGrowth[T] {
	return &FPGrowth[T]{
		core: NewCoreWithCount(minSupportCount),
	}
}

func (fp *FPGrowth[T]) Mine(transactions [][]T) []Pattern[T] {
	if len(transactions) == 0 {
		return nil
	}

	// 1. Map T -> int
	itemToID := make(map[T]int)
	idToItem := make(map[int]T)
	nextID := 1 // Start from 1, 0 is reserved for root

	coreTransactions := make([][]int, len(transactions))
	for i, tx := range transactions {
		coreTx := make([]int, len(tx))
		for j, item := range tx {
			if id, exists := itemToID[item]; exists {
				coreTx[j] = id
			} else {
				id = nextID
				nextID++
				itemToID[item] = id
				idToItem[id] = item
				coreTx[j] = id
			}
		}
		coreTransactions[i] = coreTx
	}

	// 2. Call core
	corePatterns := fp.core.Mine(coreTransactions)

	// 3. Map back
	patterns := make([]Pattern[T], len(corePatterns))
	for i, p := range corePatterns {
		items := make([]T, len(p.Items))
		for j, id := range p.Items {
			items[j] = idToItem[id]
		}
		patterns[i] = Pattern[T]{
			Items:   items,
			Support: float64(p.Support),
		}
	}

	return patterns
}
